### ISSUE TITLE
WaveOutEvent audio out of sync with CurrentTime?

### ISSUE THREAD

**Issue by jfrank14** [naudio/NAudio#489]
In my application, I use a `WaveOutEvent` to play a wav file and I have a DispatcherTimer fire every 100ms to get the current location (actually, CurrentTime) and I use this to draw a playhead over a waveform rendering of the file. When I do this, there is a slight but unacceptable latency: the playhead is always about 1/2 second ahead of the playback. I also notice that when I start playing, the CurrentTime always jumps right from zero to 0.4s. It is not possible to check the value and get 0.2s, for example. These facts seem related, but I'm not sure how.

Also `WaveOutEvent` seems to have resolution only down to 100ms; querying any more often than that shows a CurrentTime that's the same as the previous value. This is mostly okay, but when zoomed in to the second level, the play is staccato, because 100ms covers more than one pixel. Is there any way to have a higher resolution on the playback?

**Comment by ImBunky** [comment]
What size buffer are you using, pretty sure WaveOutEvent fires every time a buffer is emptied. Also, I think default latency is 300 ms. So that plus 100 ms buffer is giving you your 0.4s.

**Comment by markheath** [comment]
Yes, @ImBunky is right. You can use the `IWavePosition` interface methods to find out a more accurate representation of how many bytes of audio have been played from the device, but you need to do your own mapping onto the actual position in the file.

**Comment by simplifirepyroneil** [comment]
I am playing with getting lower latency to match the time in the playback to the actual time. This seems to help        

 waveOutDevice = new WaveOutEvent()
         {
            DesiredLatency = 10, 
            NumberOfBuffers = 200
         };

**Comment by markheath** [comment]
Closing this old issue with a little more information for anyone interesting:

`CurrentTime` on the reader/stream reflects how much audio has been **read out of the source to fill the output buffers** — not how much has actually reached the speakers. NAudio keeps `BufferMilliseconds × NumberOfBuffers` of audio queued ahead (by default 100ms × 2 = 200ms), plus the driver's own buffering, so the read position naturally runs ahead of audible playback. That's also why it advances in ~100ms steps: the source position only moves when a buffer is filled, so its granularity is one `BufferMilliseconds`.

For an accurate playhead, query what's actually been played instead:

```csharp
long bytes = waveOut.GetPosition();          // calls waveOutGetPosition directly
double seconds = (double)bytes / waveOut.OutputWaveFormat.AverageBytesPerSecond;
```

That reflects the rendered position (no half-second lead) and gives much finer resolution than the 100ms you get from the source read position.

On tuning: lowering `BufferMilliseconds` (e.g. 25–50ms) with a small `NumberOfBuffers` (2–3) reduces latency and makes the playhead smoother — that part of the earlier suggestion is sound. But **don't** push `NumberOfBuffers` up to 200: buffer *count* is the wrong lever. More buffers just means more audio queued ahead, which *increases* total latency and makes the lead worse. The low-latency lever is *smaller* buffers, not *more* of them.

Finally, if you're on (or moving to) NAudio 3: the new **`WasapiPlayer`** (built via `WasapiPlayerBuilder`) implements `IWavePosition` with a `GetPosition()` backed by the WASAPI audio clock, so you get accurate, fine-grained playback position out of the box, plus lower latency than `WaveOut`. For a waveform playhead that's the path I'd recommend for new code.
